esp-config/src/generate.rs

```diff
@@ -0,0 +1,211 @@
+                _ => return Err(ParseError(format!("Invalid boolean value: {}", s))),
+            },
+            Value::Number(_) => Value::Number(
+                match s.as_bytes() {
+                    [b'0', b'x', ..] => usize::from_str_radix(&s[2..], 16),
+                    [b'0', b'o', ..] => usize::from_str_radix(&s[2..], 8),
+                    [b'0', b'b', ..] => usize::from_str_radix(&s[2..], 2),
+                    _ => usize::from_str_radix(&s, 10),
+                }
+                .map_err(|_| ParseError(format!("Invalid numerical value: {}", s)))?,
+            ),
+            Value::String(_) => Value::String(String::from(s)),
+        };
+        Ok(())
+    }
+
+    fn as_string(&self) -> String {
+        match self {
+            Value::Bool(value) => String::from(if *value { "true" } else { "false" }),
+            Value::Number(value) => format!("{}", value),
+            Value::String(value) => value.clone(),
+        }
+    }
+}
+
+/// This function will parse any screaming snake case environment variables that
+/// match the given prefix. It will then attempt to parse the [`Value`]. Once
+/// the config has been parsed, this function will emit snake case cfg's
+/// _without_ the prefix which can be used in the dependant crate. After that,
+/// it will create a markdown table in the `OUT_DIR` under the name
+/// `{prefix}_config_table.md` where prefix has also been converted
+/// to snake case. This can be included in crate documentation to outline the
+/// available configuration options for the crate.
+///
+/// The tuple ordering for the`config` array is key, default, description.
+///
+/// Unknown keys with the supplied prefix will cause this function to panic.
+pub fn generate_config(prefix: &str, config: &[(&str, Value, &str)]) {
+    let mut prefix = screaming_snake_case(prefix);
+
+    let mut doc_table = String::from(DOC_TABLE_HEADER);
+    for (key, default, desc) in config {
+        let key = screaming_snake_case(key);
+        let default = default.as_string();
+        writeln!(doc_table, "|**{prefix}_{key}**|{desc}|{default}|").unwrap();
+    }
+
+    // only rebuild if build.rs changed. Otherwise Cargo will rebuild if any
+    // other file changed.
+    println!("cargo:rerun-if-changed=build.rs");
+
+    // Generate the template for the config
+    let mut configs = HashMap::new();
+    for (name, default, _) in config {
+        configs.insert(snake_case(name), default.clone());
+
+        // Rebuild if config envvar changed.
+        // Note this is currently buggy and requires a clean build - PR is pending https://github.com/rust-lang/cargo/pull/14058
+        println!("cargo:rerun-if-env-changed={name}");
+    }
+
+    let mut unknown = Vec::new();
+    let mut failed = Vec::new();
+
+    // Try and capture input from the environment
+    for (var, value) in env::vars() {
+        if let Some(name) = var.strip_prefix(&format!("{prefix}_")) {
```

Comment:
This line will only accept SCREAMING_SNAKE_CASE prefixes, while I believe we could be a bit more lenient to what we accept from the user.

There's quite a bit of case conversion happening in this script all over the place and I think that can become somewhat hard to reason about. My advice would be to just SSC everything as soon as possible and don't bother with lower-case snake case, except when writing the output.

Reply:
I've pushed a commit which reduces the number of case conversions, you were right it was a bit unnecessary.

> while I believe we could be a bit more lenient to what we accept from the user

I think it's best to be stricter here, especially since we bomb out if we find any unknown variable. We can relax this later on if we get enough requests for it, but IMO we're not asking for too much in the first place here.